I think something may be missing. I get build errors (installed `xxHash`).
The function `avcodec_parameters_from_context` does not seem to be defined anywhere.
I strongly suspect something isn't installed -- what defines `AVStream`?

```
/home/rvn-ros/Documents/RavenOps/rviz_ws/src/rviz/src/rviz_video_encoder/video_encoder.c: In function ‘video_encoder_init’:
/home/rvn-ros/Documents/RavenOps/rviz_ws/src/rviz/src/rviz_video_encoder/video_encoder.c:160:11: warning: implicit declaration of function ‘avcodec_parameters_from_context’ [-Wimplicit-function-declaration]
     ret = avcodec_parameters_from_context(enc->st->codecpar, enc->cc);
           ^
/home/rvn-ros/Documents/RavenOps/rviz_ws/src/rviz/src/rviz_video_encoder/video_encoder.c:160:50: error: ‘AVStream {aka struct AVStream}’ has no member named ‘codecpar’
     ret = avcodec_parameters_from_context(enc->st->codecpar, enc->cc);
                                                  ^
/home/rvn-ros/Documents/RavenOps/rviz_ws/src/rviz/src/rviz_video_encoder/video_encoder.c: In function ‘video_encoder_send_frame’:
/home/rvn-ros/Documents/RavenOps/rviz_ws/src/rviz/src/rviz_video_encoder/video_encoder.c:191:11: warning: implicit declaration of function ‘avcodec_send_frame’ [-Wimplicit-function-declaration]
     ret = avcodec_send_frame(enc->cc,frame);
           ^
/home/rvn-ros/Documents/RavenOps/rviz_ws/src/rviz/src/rviz_video_encoder/video_encoder.c:202:15: warning: implicit declaration of function ‘avcodec_receive_packet’ [-Wimplicit-function-declaration]
         ret = avcodec_receive_packet(enc->cc,&enc->pbuf);
               ^
rviz/src/rviz_video_encoder/CMakeFiles/rviz_video_encoder.dir/build.make:62: recipe for target 'rviz/src/rviz_video_encoder/CMakeFiles/rviz_video_encoder.dir/video_encoder.c.o' failed
make[2]: *** [rviz/src/rviz_video_encoder/CMakeFiles/rviz_video_encoder.dir/video_encoder.c.o] Error 1
CMakeFiles/Makefile2:3047: recipe for target 'rviz/src/rviz_video_encoder/CMakeFiles/rviz_video_encoder.dir/all' failed
make[1]: *** [rviz/src/rviz_video_encoder/CMakeFiles/rviz_video_encoder.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
[  9%] Built target rviz_automoc
Makefile:138: recipe for target 'all' failed
```

--- a/src/rviz/visualization_manager.cpp
+++ b/src/rviz/visualization_manager.cpp
@@ -371,8 +371,7 @@ ros::CallbackQueueInterface* VisualizationManager::getUpdateQueue()
 
 void VisualizationManager::startUpdate()
 {
-    //  float interval = 1000.0 / float(fps_property_->getInt());
-  float interval = 0.0;
+  float interval = 1000.0 / float(fps_property_->getInt());
   update_timer_->start( interval );
 }
 

--- a/src/rviz/visualization_manager.cpp
+++ b/src/rviz/visualization_manager.cpp
@@ -519,6 +519,12 @@ void VisualizationManager::onUpdate()
     if (shouldDump  && dump_images_config_->bagDuration > 0)
     {
       QPixmap screenshot_ = screen_->grabWindow(window_->winId());
+      if (screenshot_.width() % 2 || screenshot_.height() % 2) {
+          // ensure screenshot width are height are divisible by two
+          int w = screenshot_.width();
+          int h = screenshot_.height();
+          screenshot_ = screenshot_.copy(0, 0, w - (w%2), h - (h%2));
+      }
       QImage img = screenshot_.toImage();
 
       if ( private_->venc_ == NULL){
@@ -594,7 +600,7 @@ void VisualizationManager::onUpdate()
           exit(EXIT_FAILURE);
       }
 
-      // screenshot
+      // thumbnail
       double midway = dump_images_config_->bagDuration / 2.0;
       double curTime = dump_images_config_->lastEventTime - private_->bag_start;
       if ( curTime >= midway && !private_->thumbnail_taken )
@@ -606,7 +612,7 @@ void VisualizationManager::onUpdate()
           QImageWriter writer(filename);
           if (!writer.write(img.scaledToWidth(dump_images_config_->thumbWidth)))
           {
-              ROS_ERROR("failed writing screenshot file %s: err %d",dump_images_config_->thumb_path.c_str(),writer.error());
+              ROS_ERROR("failed writing thumbnail file %s: err %d",dump_images_config_->thumb_path.c_str(),writer.error());
               exit(EXIT_FAILURE);
           }
           ROS_INFO("Wrote thumbnail at %f of %f (midway = %f)\n", curTime, dump_images_config_->bagDuration, midway);

--- a/src/rviz/visualization_manager.cpp
+++ b/src/rviz/visualization_manager.cpp
@@ -157,7 +157,7 @@ VisualizationManager::VisualizationManager(RenderPanel* render_panel,DumpImagesC
 , frame_count_(0)
 , dumped_frame_count_(0)
 , window_manager_(wm)
-, private_( new VisualizationManagerPrivate )
+, private_( new VisualizationManagerPrivate {})
 {
   // visibility_bit_allocator_ is listed after default_visibility_bit_ (and thus initialized later be default):
   default_visibility_bit_ = visibility_bit_allocator_.allocBit();
@@ -519,11 +519,11 @@ void VisualizationManager::onUpdate()
     if (shouldDump  && dump_images_config_->bagDuration > 0)
     {
       QPixmap screenshot_ = screen_->grabWindow(window_->winId());
-      if (screenshot_.width() % 2 || screenshot_.height() % 2) {
-          // ensure screenshot width are height are divisible by two
+      if (screenshot_.width() % 4 || screenshot_.height() % 4) {
+          // ensure screenshot width are height are divisible by four
           int w = screenshot_.width();
           int h = screenshot_.height();
-          screenshot_ = screenshot_.copy(0, 0, w - (w%2), h - (h%2));
+          screenshot_ = screenshot_.copy(0, 0, w - (w % 4), h - (h % 4));
       }
       QImage img = screenshot_.toImage();
 

--- a/src/rviz_video_encoder/video_encoder.c
+++ b/src/rviz_video_encoder/video_encoder.c
@@ -46,19 +46,21 @@ VideoEncoder* video_encoder_init(VideoEncodeParams params){
     memset(enc,0,sizeof(VideoEncoder));
 
     enc->params = params;
-
-    printf("[video encoder] capping width at %dpx\n",enc->params.maxWidth);
-    float aspectRatio = (float) enc->params.width / (float) enc->params.height;
-    enc->out_width = enc->params.width > enc->params.maxWidth ? enc->params.maxWidth : enc->params.width;
-    enc->out_height = (int) ((float) enc->out_width / aspectRatio);
-
-    // ensure output dimensions are divisible by 2
-    if (enc->out_width % 2) {
-        enc->out_width--;
+    if (enc->params.width % 2 || enc->params.height % 2){
+        fprintf(stderr,"%dx%d, dimensions not divisible by two!\n",enc->params.width,enc->params.height);
+        return NULL;
     }
-    if (enc->out_height % 2) {
-        enc->out_height--;
+    if (enc->params.width > enc->params.maxWidth) {
+        printf("[video encoder] WARN max width exceeded, scaling down to max width %dpx\n",enc->params.maxWidth);
+        float aspectRatio = (float) enc->params.width / (float) enc->params.height;
+        enc->out_width = enc->params.maxWidth;
+        enc->out_height = (int)( (float) enc->out_width / aspectRatio);
+    }else{
+        enc->out_width = enc->params.width;
+        enc->out_height = enc->params.height;
     }
+    enc->out_width -= enc->out_width % 2;
+    enc->out_height -= enc->out_height % 2;
 
     printf("[video encoder init] %d x %d --> %d x %d @ %f FPS\n",
 	   enc->params.width,enc->params.height,
@@ -75,16 +77,15 @@ VideoEncoder* video_encoder_init(VideoEncodeParams params){
     enc->frame_raw->width = enc->params.width;
     enc->frame_raw->height = enc->params.height;
     enc->frame_raw->format = enc->params.inFmt;
-
     int ret;
-    ret = av_frame_get_buffer(enc->frame_scaled, 32);
+    ret = av_frame_get_buffer(enc->frame_scaled,32);
     if (ret < 0) {
-        fprintf(stderr, "Could not allocate frame data: %s\n",av_err2str(ret));
+        fprintf(stderr, "Could not allocate scaled frame data: %s\n",av_err2str(ret));
         return NULL;
     }
-    ret = av_frame_get_buffer(enc->frame_raw, 32);
+    ret = av_frame_get_buffer(enc->frame_raw,32);
     if (ret < 0) {
-        fprintf(stderr, "Could not allocate frame data: %s\n",av_err2str(ret));
+        fprintf(stderr, "Could not allocate raw frame data: %s\n",av_err2str(ret));
         return NULL;
     }
 
@@ -223,25 +224,14 @@ static int video_encoder_send_frame(VideoEncoder *enc, AVFrame *frame){
 }
 
 int video_encoder_encode_frame(VideoEncoder *enc, const uint8_t *pixels){
-    int ret;
-    ret = av_frame_make_writable(enc->frame_raw);
-    if(ret < 0 ){
-        fprintf(stderr, "[video encoder] error allocating frame: %s\n",av_err2str(ret));
-        return ret;
-    }
-    ret = av_frame_make_writable(enc->frame_scaled);
-    if(ret < 0 ){
-        fprintf(stderr, "[video encoder] error allocating frame: %s\n",av_err2str(ret));
-        return ret;
-    }
 
     int bytesFilled = av_image_fill_arrays(enc->frame_raw->data,
                                            enc->frame_raw->linesize,
                                            pixels,
                                            enc->frame_raw->format,
                                            enc->frame_raw->width,enc->frame_raw->height,32);
-    if(!bytesFilled){
-       fprintf(stderr,"Could not fill raw picture bytes\n");
+    if(bytesFilled < 0){
+       fprintf(stderr,"Could not fill raw picture bytes: %s\n",av_err2str(bytesFilled));
        return -1;
     }
 
@@ -257,7 +247,7 @@ int video_encoder_encode_frame(VideoEncoder *enc, const uint8_t *pixels){
     }
 
     enc->frame_scaled->pts = enc->num_frames++;
-    sws_scale(enc->sws,
+    int ret = sws_scale(enc->sws,
               (const uint8_t * const *) enc->frame_raw->data,
 	      enc->frame_raw->linesize,
               0,
@@ -265,9 +255,10 @@ int video_encoder_encode_frame(VideoEncoder *enc, const uint8_t *pixels){
               enc->frame_scaled->data,
               enc->frame_scaled->linesize
               );
-
-
-
+    if (ret < 0){
+        fprintf(stderr,"failed to scale frame: %s\n",av_err2str(ret));
+        return ret;
+    }
     return video_encoder_send_frame(enc,enc->frame_scaled);
 }
 

